flax/linen/module.py

```diff
@@ -1104,6 +1104,17 @@ def other_fn(instance, ...):
     if method is None:
       method = self.__call__
     method = _get_unbound_fn(method)
+    # Try to detect if user accidentally passed {'params': {'params': ...} or
+    # {'kernel': ...} as `variables`.
+    if 'params' in variables and isinstance(
+        variables['params'], FrozenDict) and 'params' in variables['params']:
+      raise ValueError(
```

Comment:
I'm thinking it may be nice to add a new error class to errors.py: https://github.com/google/flax/blob/main/flax/errors.py

Then we can add a new class covering both cases, and we explain how a variable dict should look like and what mistakes users typically make (i.e., the two mistakes we find here).

We could also add a separate error class for each case, that may be even more clear.

WDYT?

Reply:
Happy to add a new error class to cover badly structured variable dicts, including explanation as you mentioned.
I think it might be overkill to make a separate error class per case, given that these two cases are really instances of the same kind of user mistake, which means we would have to (more or less) duplicate the explanation text for those two and thus increase the maintenance burden.